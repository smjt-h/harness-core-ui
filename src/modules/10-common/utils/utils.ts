/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export const isCommunityPlan = (): boolean => false

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined
