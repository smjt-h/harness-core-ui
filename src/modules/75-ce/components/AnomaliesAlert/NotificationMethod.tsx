/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import {
  Layout,
  StepProps,
  Color,
  FontVariation,
  Text,
  ButtonVariation,
  Button,
  Icon,
  FormInput,
  SelectOption
} from '@harness/uicore'
import { FieldArray, FormikProps } from 'formik'
import { defaultTo } from 'lodash-es'
import { useStrings } from 'framework/strings'
import { channels, notificationChannelsList } from '@ce/constants'
import type { StringsMap } from 'stringTypes'
import css from './AnomaliesAlertDialog.module.scss'

interface StepData {
  name: string
}

interface NotificationChannelProps {
  name: string
  onClose: () => void
  formikProps: FormikProps<NotificationValues>
}

interface Alerts {
  channelName?: channels
  channelUrl?: string | string[]
}

export interface NotificationValues {
  perspective: string
  alertList: Alerts[]
}

const NotificationMethod: React.FC<StepProps<StepData> & NotificationChannelProps> = props => {
  const { getString } = useStrings()
  const notificationList = defaultTo(props.formikProps?.values?.alertList, [])
  const channelList = notificationChannelsList.map(item => {
    return {
      ...item,
      label: getString(item.label as keyof StringsMap)
    }
  })

  return (
    <React.Fragment>
      <Icon name="cross" size={16} onClick={() => props.onClose()} className={css.closeBtn} data-testId="crossModal" />
      <Layout.Vertical spacing="large">
        <Text color={Color.BLACK} font={{ variation: FontVariation.H5 }}>
          {props.name}
        </Text>
        <Text color={Color.GREY_500} font={{ variation: FontVariation.SMALL }}>
          {getString('ce.anomalyDetection.notificationAlerts.notificationStepSubtext')}
        </Text>
      </Layout.Vertical>
      <Layout.Vertical spacing="medium" className={css.addChannelWrapper}>
        <Text color={Color.BLACK} font={{ variation: FontVariation.H6 }}>
          {getString('ce.anomalyDetection.notificationAlerts.alertChannelHeading')}
        </Text>
        <FieldArray
          name="alertList"
          render={arrayHelpers => {
            const removeRow = (index: number) => {
              arrayHelpers.remove(index)
            }

            return (
              <div className={css.alertChannelWrapper}>
                {notificationList.map((_, index) => (
                  <Layout.Horizontal key={index} spacing="xlarge" className={css.addAlertChannel}>
                    <FormInput.Select
                      items={channelList as SelectOption[]}
                      key={`channelName_${index}`}
                      name={`alertList.${index}.channelName`}
                      className={css.channelSelection}
                      placeholder={getString('ce.anomalyDetection.notificationAlerts.selectChannelPlaceholder')}
                      data-testid={`notification-channel-${index}`}
                    />
                    <FormInput.Text
                      name={`alertList.${index}.channelUrl`}
                      key={`channelUrl_${index}`}
                      placeholder={getString('ce.anomalyDetection.notificationAlerts.urlInputPlaceholder')}
                      className={css.urlInput}
                    />
                    <Icon name="main-trash" size={16} onClick={() => removeRow(index)} data-testid="deleteChannel" />
                  </Layout.Horizontal>
                ))}
                <Button
                  text={getString('ce.anomalyDetection.notificationAlerts.addChannelBtn')}
                  icon="plus"
                  onClick={() => arrayHelpers.push({ channelName: '', channelUrl: '' })}
                  variation={ButtonVariation.LINK}
                  className={css.addChannelBtn}
                />
              </div>
            )
          }}
        />
      </Layout.Vertical>
      <Layout.Horizontal className={css.actionBtn} spacing="medium">
        <Button
          onClick={() => props.previousStep?.({ name: props.name || '' })}
          text={getString('back')}
          icon="chevron-left"
          variation={ButtonVariation.TERTIARY}
        />
        <Button
          type="submit"
          variation={ButtonVariation.PRIMARY}
          rightIcon="chevron-right"
          onClick={() => props.nextStep?.({ name: props.name || '' })}
          text={getString('saveAndContinue')}
          data-testid="submitForm"
          disabled={!notificationList.length}
        />
      </Layout.Horizontal>
    </React.Fragment>
  )
}

export default NotificationMethod
