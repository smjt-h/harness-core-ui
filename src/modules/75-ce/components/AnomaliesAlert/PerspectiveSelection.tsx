/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import {
  Layout,
  StepProps,
  Color,
  FontVariation,
  Text,
  ButtonVariation,
  Button,
  Icon,
  FormInput,
  SelectOption
} from '@harness/uicore'
import type { FormikProps } from 'formik'
import { useStrings } from 'framework/strings'
import type { channels } from '@ce/constants'
import css from './AnomaliesAlertDialog.module.scss'

interface AlertsOverviewProps {
  name: string
  onClose: () => void
  items: SelectOption[]
  formikProps: FormikProps<NotificationValues>
}

interface StepData {
  name: string
}

interface Alerts {
  channelName?: channels
  channelUrl?: string | string[]
}

interface NotificationValues {
  perspective: string
  alertList: Alerts[]
}

const PerspectiveSelection: React.FC<StepProps<StepData> & AlertsOverviewProps> = props => {
  const { getString } = useStrings()

  return (
    <React.Fragment>
      <Icon name="cross" size={16} onClick={() => props.onClose()} className={css.closeBtn} />
      <Layout.Vertical spacing="xxlarge">
        <Text color={Color.BLACK} font={{ variation: FontVariation.H5 }}>
          {props.name}
        </Text>
        <FormInput.Select
          items={props.items}
          className={css.perspectiveSelection}
          name={'perspective'}
          label={getString('ce.anomalyDetection.notificationAlerts.selectPerspective')}
          placeholder={getString('ce.anomalyDetection.notificationAlerts.selectPerspective')}
        />
      </Layout.Vertical>
      <Button
        className={css.actionBtn}
        rightIcon="chevron-right"
        variation={ButtonVariation.PRIMARY}
        onClick={() => props.nextStep?.({ name: props.name || '' })}
        text={getString('saveAndContinue')}
        disabled={!props.formikProps.values.perspective}
      />
    </React.Fragment>
  )
}

export default PerspectiveSelection
