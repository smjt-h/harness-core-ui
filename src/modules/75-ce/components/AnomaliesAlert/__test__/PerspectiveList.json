{
  "data": {
    "perspectives": {
      "customerViews": [
        { "id": "NBR8rYCgTLyhbKgwxnjKUw", "name": "NBR8rYCgTLyhbKgwxnjKUw", "__typename": "QLCEView" },
        { "id": "Yv-hfTBQTc2LO6U-rzZvBA", "name": "Yv-hfTBQTc2LO6U-rzZvBA", "__typename": "QLCEView" },
        { "id": "uaTTAOdGQOiugVoZIG2lcA", "name": "uaTTAOdGQOiugVoZIG2lcA", "__typename": "QLCEView" },
        { "id": "e6V1JG61QWubhV89vAmUIg", "name": "e6V1JG61QWubhV89vAmUIg", "__typename": "QLCEView" }
      ],
      "__typename": "PerspectiveData"
    }
  }
}
