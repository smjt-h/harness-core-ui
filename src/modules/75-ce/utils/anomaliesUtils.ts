/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { CloudProvider } from '@ce/types'
import type { QlceViewFilterInput } from 'services/ce/services'
import { DEFAULT_GROUP_BY } from './perspectiveUtils'

const gcpEntities = ['gcpProduct', 'gcpProjectId', 'gcpSKUDescription']
const awsEntities = ['awsUsageAccountId', 'awsServiceCode', 'awsInstancetype', 'awsUsageType']
const azureEntities: string[] = []
const clusterEntities = ['clusterName', 'clusterId', 'namespace', 'workloadName', 'workloadType']

const cloudProviderToEntityMapping = {
  [CloudProvider.GCP]: gcpEntities,
  [CloudProvider.AWS]: awsEntities,
  [CloudProvider.CLUSTER]: clusterEntities,
  [CloudProvider.AZURE]: azureEntities
}

const fieldToFieldNameMapping: Record<string, string> = {
  gcpProduct: 'Product',
  gcpProjectId: 'Project',
  gcpSKUDescription: 'SKUs',
  clusterName: 'Cluster Name',
  namespace: 'Namespace',
  workloadName: 'Workload',
  awsUsageAccountId: 'Account',
  awsServiceCode: 'Service',
  awsInstancetype: 'Instance Type',
  awsUsageType: 'Usage Type',
  workloadType: 'Workload Type',
  gcpSkuDescription: 'SKUs'
}

export function generateFilters(
  entityMap: Record<string, string>,
  cloudProvider: CloudProvider
): QlceViewFilterInput[] {
  const filters: any = []
  const relatedEntities = cloudProviderToEntityMapping[cloudProvider]
  relatedEntities.length &&
    relatedEntities.forEach(entity => {
      if (entityMap[entity]) {
        filters.push({
          field: {
            fieldId: entity,
            fieldName: fieldToFieldNameMapping[entity] || '',
            identifier: cloudProvider,
            identifierName: cloudProvider
          },
          operator: 'IN',
          type: 'VIEW_ID_CONDITION',
          values: [entityMap[entity]]
        })
      }
    })

  return filters
}

export function generateGroupBy(field: string, cloudProvider: CloudProvider): any {
  return field
    ? {
        fieldId: field,
        fieldName: fieldToFieldNameMapping[field] || '',
        identifier: cloudProvider,
        identifierName: cloudProvider
      }
    : DEFAULT_GROUP_BY
}

export function getCloudProviderFromFields(entityMap: Record<string, string | null>): CloudProvider {
  if (entityMap.gcpProjectId) {
    return CloudProvider.GCP
  }
  if (entityMap.awsUsageAccountId) {
    return CloudProvider.AWS
  }
  return CloudProvider.CLUSTER
}

export function getFiltersFromEnityMap(
  entityMap: Array<Record<string, string>>,
  cloudProvider: CloudProvider
): QlceViewFilterInput[] {
  const filters: any = []
  const relatedEntities = cloudProviderToEntityMapping[cloudProvider]

  relatedEntities.length &&
    relatedEntities.forEach(entity => {
      if (entityMap.length && entityMap[0][entity]) {
        const values: string[] = []
        entityMap.forEach(obj => {
          obj[entity] && values.push(obj[entity])
        })

        values.length &&
          filters.push({
            field: {
              fieldId: entity,
              fieldName: fieldToFieldNameMapping[entity] || '',
              identifier: cloudProvider,
              identifierName: cloudProvider
            },
            operator: 'IN',
            type: 'VIEW_ID_CONDITION',
            values: values
          })
      }
    })

  return filters
}
