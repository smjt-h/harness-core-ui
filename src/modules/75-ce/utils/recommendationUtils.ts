/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import type { IState } from '@ce/components/NodeRecommendation/constants'
import type { RecommendClusterRequest } from 'services/ce/recommenderService'

export const getProviderIcon = (provider: string): IconName => {
  const iconMapping: Record<string, IconName> = {
    google: 'gcp',
    azure: 'service-azure',
    amazon: 'service-aws'
  }

  return iconMapping[provider] || 'app-kubernetes'
}

export const addBufferToValue = (value: number, bufferPercentage: number, precision?: number): number =>
  +(((100 + bufferPercentage) / 100) * value).toFixed(precision || 2)

export const calculateSavingsPercentage = (savings: number, totalCost: number): string =>
  `(${Math.floor((savings / totalCost) * 100)}%)`

export const isResourceConsistent = (
  sumCpu: number,
  sumMemory: number,
  maxCpu: number,
  maxMemory: number,
  buffer: number
): boolean => {
  const sumCpuWithBuffer = addBufferToValue(sumCpu, buffer)
  const sumMemWithBuffer = addBufferToValue(sumMemory, buffer)

  const isInconsistent =
    Math.round(sumCpuWithBuffer) < Math.round(maxCpu) || Math.round(sumMemWithBuffer) < Math.round(maxMemory)

  const anyZero =
    Math.round(sumCpuWithBuffer) === 0 ||
    Math.round(sumMemWithBuffer) === 0 ||
    Math.round(maxCpu) === 0 ||
    Math.round(maxMemory) === 0

  return !isInconsistent && !anyZero
}

export const calculateNodes = (
  sumCpu: number,
  sumMemory: number,
  maxCpu: number,
  maxMemory: number,
  minNodes: number
): { maximumNodes: number; minimumNodes: number } => {
  let minimumNodes = minNodes || 3

  let maximumNodes = Math.min(Math.floor(sumCpu / maxCpu), Math.floor(sumMemory / maxMemory))
  maximumNodes = Math.max(maximumNodes, 1)

  if (maximumNodes < minNodes) {
    minimumNodes = maximumNodes
  }

  return { maximumNodes, minimumNodes }
}

export const convertStateToRecommendClusterPayload = (
  state: IState,
  resourceRequirement: RecommendClusterRequest,
  buffer: number
): RecommendClusterRequest => {
  const sumCpuWithBuffer = addBufferToValue(state.sumCpu, buffer)
  const sumMemWithBuffer = addBufferToValue(state.sumMem, buffer)

  const { maximumNodes, minimumNodes } = calculateNodes(
    sumCpuWithBuffer,
    sumMemWithBuffer,
    state.maxCpu,
    +state.maxMemory,
    +state.minNodes
  )

  return {
    ...resourceRequirement,
    sumCpu: +sumCpuWithBuffer,
    sumMem: +sumMemWithBuffer,
    minCpu: +state.maxCpu,
    minMem: +state.maxMemory,
    maxNodes: maximumNodes,
    minNodes: minimumNodes,
    includeSeries: state.includeSeries,
    includeTypes: state.includeTypes,
    excludeSeries: state.excludeSeries,
    excludeTypes: state.excludeTypes
  }
}

export const addBufferToState = (state: IState, buffer: number): IState => ({
  ...state,
  sumCpu: addBufferToValue(state.sumCpu, buffer),
  sumMem: addBufferToValue(state.sumMem, buffer)
})

interface InstanceFamilies {
  includeSeries: string[]
  includeTypes: string[]
  excludeSeries: string[]
  excludeTypes: string[]
}

export const getInstanceFamiliesFromState = (state: IState): InstanceFamilies => ({
  includeSeries: state.includeSeries,
  includeTypes: state.includeTypes,
  excludeSeries: state.excludeSeries,
  excludeTypes: state.excludeTypes
})
