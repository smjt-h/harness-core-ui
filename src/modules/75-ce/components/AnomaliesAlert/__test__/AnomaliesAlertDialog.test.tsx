/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { act, fireEvent, render, waitFor } from '@testing-library/react'
import React from 'react'
import { fromValue } from 'wonka'
import type { DocumentNode } from 'graphql'
import { Provider } from 'urql'
import { Dialog, Formik, StepWizard } from '@harness/uicore'
import type { IDialogProps } from '@blueprintjs/core'
import { defaultTo } from 'lodash-es'
import { findDialogContainer, TestWrapper } from '@common/utils/testUtils'
import { FetchPerspectiveListDocument } from 'services/ce/services'
import type { CCMPerspectiveNotificationChannelsDTO } from 'services/ce'
import PerspectiveList from './PerspectiveList.json'
import PerspectiveSelection from '../PerspectiveSelection'
import NotificationMethod, { NotificationValues } from '../NotificationMethod'
import useAnomaliesAlertDialog from '../AnomaliesAlertDialog'

const selectedAlert: CCMPerspectiveNotificationChannelsDTO = {
  perspectiveId: 'perspectiveId',
  channels: [
    {
      notificationChannelType: 'SLACK',
      channelUrls: []
    }
  ]
}

const params = {
  accountId: 'TEST_ACC'
}

const modalPropsLight: IDialogProps = {
  isOpen: true,
  enforceFocus: false,
  style: {
    width: 1100,
    position: 'relative',
    minHeight: 600,
    borderLeft: 0,
    paddingBottom: 0,
    overflow: 'hidden'
  }
}

const handleSubmit = jest.fn().mockImplementation(async () => {
  try {
    await Promise.resolve(true)
  } catch (err) {
    expect(err).toBeTruthy()
  }
})

const hideAnomaliesAlertModal = jest.fn()

jest.mock('services/ce', () => {
  return {
    useCreateNotificationSetting: jest.fn().mockImplementation(() => ({
      mutate: async () => {
        return {
          status: 'SUCCESS',
          data: {}
        }
      }
    })),
    useUpdateNotificationSetting: jest.fn().mockImplementation(() => ({
      mutate: async () => {
        return {
          status: 'SUCCESS',
          data: {}
        }
      }
    }))
  }
})

const Wrapped = (): React.ReactElement => {
  const { openAnomaliesAlertModal } = useAnomaliesAlertDialog({
    setRefetchingState: jest.fn(),
    selectedAlert: selectedAlert
  })
  const onBtnClick = () => {
    openAnomaliesAlertModal()
  }

  return (
    <>
      <button className="opnModal" onClick={onBtnClick} />
      <Dialog {...modalPropsLight} onClose={hideAnomaliesAlertModal}>
        <Formik<NotificationValues>
          initialValues={{
            perspective: defaultTo(selectedAlert.perspectiveId, ''),
            alertList: []
          }}
          formName={'createNotificationAlert'}
          onSubmit={data => handleSubmit(data)}
        >
          {formikProps => {
            return (
              <StepWizard>
                <PerspectiveSelection
                  name={''}
                  onClose={hideAnomaliesAlertModal}
                  items={[]}
                  formikProps={formikProps}
                />
                <NotificationMethod name={''} onClose={hideAnomaliesAlertModal} formikProps={formikProps} />
              </StepWizard>
            )
          }}
        </Formik>
      </Dialog>
    </>
  )
}

describe('Test case for anomalies alert dialog', () => {
  test('should work as expected', async () => {
    const responseState = {
      executeQuery: ({ query }: { query: DocumentNode }) => {
        if (query === FetchPerspectiveListDocument) {
          return fromValue(PerspectiveList)
        }
        return fromValue({})
      }
    }
    const { container, getByText, getAllByText } = render(
      <Provider value={responseState as any}>
        <TestWrapper>
          <Wrapped />
        </TestWrapper>
      </Provider>
    )

    const openModal = container.querySelector('.opnModal')
    await act(async () => {
      fireEvent.click(openModal!)
    })

    const modal = findDialogContainer()
    expect(modal).toBeDefined()

    expect(getAllByText('ce.anomalyDetection.notificationAlerts.selectPerspective')).toBeDefined()

    const saveAndContinueBtn = getAllByText('saveAndContinue')[0]
    act(() => {
      fireEvent.click(saveAndContinueBtn!)
    })
    expect(getByText('ce.anomalyDetection.notificationAlerts.notificationStepSubtext')).toBeDefined()

    const submitFormBtn = modal?.querySelector('[data-testid="submitForm"]')
    expect(submitFormBtn).toBeDefined()
    act(() => {
      fireEvent.click(submitFormBtn!)
    })

    expect(modal).toMatchSnapshot()
  })

  test('Should be able to close the dialog on close', async () => {
    const responseState = {
      executeQuery: ({ query }: { query: DocumentNode }) => {
        if (query === FetchPerspectiveListDocument) {
          return fromValue(PerspectiveList)
        }
        return fromValue({})
      }
    }

    render(
      <TestWrapper pathParams={params}>
        <Provider value={responseState as any}>
          <Wrapped />
        </Provider>
      </TestWrapper>
    )

    const modal = findDialogContainer()
    expect(modal).toBeDefined()

    await waitFor(() => Promise.resolve())

    const crossIcon = modal?.querySelector('.closeBtn')
    act(() => {
      fireEvent.click(crossIcon!)
    })

    expect(hideAnomaliesAlertModal).toBeCalled()
  })
})
