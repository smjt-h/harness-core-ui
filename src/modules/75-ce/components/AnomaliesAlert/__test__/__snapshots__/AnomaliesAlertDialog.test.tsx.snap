// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for anomalies alert dialog should work as expected 1`] = `
<div
  class="bp3-dialog Dialog--dialog"
  style="width: 1100px; position: relative; min-height: 600px; border-left: 0; padding-bottom: 0px; overflow: hidden;"
>
  <span
    class="bp3-icon StyledProps--main Dialog--close"
    data-icon="Stroke"
  >
    <svg
      fill="none"
      height="15"
      width="15"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 .8l6 6 6-6-12 12 6-6 6 6"
        stroke="#8F90A6"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </span>
  <div
    class="StyledProps--font StyledProps--main Dialog--children StyledProps--padding-none"
  >
    <div
      class="OverlaySpinner--overlaySpinner"
    >
      <div
        class="StepWizard--main StepWizard--navBar"
      >
        <div
          class="StepWizard--navBarList"
        >
          <div
            class="StepWizard--navStep StepWizard--completedStep"
          >
            <span
              class="StepWizard--completedIcon"
            >
              <span
                class="bp3-icon bp3-icon-small-tick StyledProps--main StyledProps--color StyledProps--color-primary7"
                icon="small-tick"
              >
                <svg
                  data-icon="small-tick"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <desc>
                    small-tick
                  </desc>
                  <path
                    d="M15 5c-.28 0-.53.11-.71.29L8 11.59l-2.29-2.3a1.003 1.003 0 00-1.42 1.42l3 3c.18.18.43.29.71.29s.53-.11.71-.29l7-7A1.003 1.003 0 0015 5z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </span>
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StepWizard--stepName"
              style="--text-line-clamp: 2; width: 200px;"
            >
              Step 1
            </p>
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StepWizard--stepName"
              style="--text-line-clamp: 2; width: 200px;"
            >
              
            </p>
          </div>
          <div
            class="StepWizard--navStep StepWizard--activeStep"
          >
            <span
              class="StepWizard--number"
            >
              2
            </span>
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StepWizard--stepName"
              style="--text-line-clamp: 2; width: 200px;"
            >
              Step 2
            </p>
          </div>
        </div>
        <div
          class="StepWizard--stepDetails"
        >
           
          <span
            class="bp3-icon bp3-icon-cross StyledProps--main closeBtn"
            data-testid="crossModal"
            icon="cross"
          >
            <svg
              data-icon="cross"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                cross
              </desc>
              <path
                d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <div
            class="Layout--vertical Layout--layout-spacing-large StyledProps--main"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--font-variation-h5"
            >
              
            </p>
            <p
              class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--font-variation-small"
            >
              ce.anomalyDetection.notificationAlerts.notificationStepSubtext
            </p>
          </div>
          <div
            class="Layout--vertical Layout--layout-spacing-medium StyledProps--main addChannelWrapper"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--font-variation-h6"
            >
              ce.anomalyDetection.notificationAlerts.alertChannelHeading
            </p>
            <div
              class="alertChannelWrapper"
            >
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main addChannelBtn Button--with-current-color Button--withLeftIcon Button--variation Button--variation-link"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
                  icon="plus"
                >
                  <svg
                    data-icon="plus"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      plus
                    </desc>
                    <path
                      d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="bp3-button-text"
                >
                  ce.anomalyDetection.notificationAlerts.addChannelBtn
                </span>
              </button>
            </div>
          </div>
          <div
            class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main actionBtn"
          >
            <button
              class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-tertiary"
              type="button"
            >
              <span
                class="bp3-icon bp3-icon-chevron-left StyledProps--main StyledProps--padding-right-xsmall"
                icon="chevron-left"
              >
                <svg
                  data-icon="chevron-left"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    chevron-left
                  </desc>
                  <path
                    d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="bp3-button-text"
              >
                back
              </span>
            </button>
            <button
              class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withRightIcon Button--variation Button--variation-primary"
              data-testid="submitForm"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              <span
                class="bp3-button-text"
              >
                saveAndContinue
              </span>
              <span
                class="bp3-icon bp3-icon-chevron-right StyledProps--main"
                icon="chevron-right"
              >
                <svg
                  data-icon="chevron-right"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    chevron-right
                  </desc>
                  <path
                    d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <span
          class="bp3-icon StyledProps--main StepWizard--harnessWatermark StyledProps--color StyledProps--color-grey50"
          data-icon="harness-with-color"
        >
          <svg
            height="346"
            viewBox="0 0 25 25"
            width="346"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22.974 8.186L16.708 1.88a7.023 7.023 0 00-2.59-1.572C11.98-.409 9.858.154 8.135 1.88L1.868 8.186a7.081 7.081 0 00-1.562 2.606c-.713 2.153-.153 4.289 1.562 6.023l6.266 6.306a7.113 7.113 0 002.589 1.58 5.663 5.663 0 001.808.308c1.503 0 2.946-.64 4.177-1.88l6.266-6.305a7.08 7.08 0 001.562-2.606c.713-2.162.153-4.298-1.562-6.032zm-10.07-4.477c.706.213 1.232.7 1.274.735l1.851 1.862-3.608 3.631-3.608-3.63 1.859-1.872c.5-.496 1.197-1.034 2.233-.726zm-9.219 8.305c.212-.71.696-1.24.73-1.282l1.85-1.863 3.609 3.632-3.608 3.63-1.86-1.87c-.492-.504-1.027-1.214-.721-2.247zm8.252 9.279c-.705-.214-1.231-.701-1.273-.735l-1.851-1.863 3.608-3.631 3.608 3.631-1.86 1.871c-.5.496-1.205 1.034-2.232.726zm9.22-8.305c-.212.709-.696 1.239-.73 1.281l-1.851 1.863-3.608-3.631 3.608-3.632 1.86 1.872c.492.504 1.027 1.204.72 2.247z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <footer />
</div>
`;
